src/main/java/io/aiven/kafka/auth/audit/Auditor.java

```diff
@@ -0,0 +1,142 @@
+/**
+ * Copyright (c) 2020 Aiven, Helsinki, Finland. https://aiven.io/
+ */
+
+package io.aiven.kafka.auth.audit;
+
+import java.io.Serializable;
+import java.net.InetAddress;
+import java.time.format.DateTimeFormatter;
+import java.util.Objects;
+import java.util.StringJoiner;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.kafka.common.security.auth.KafkaPrincipal;
+
+import kafka.network.RequestChannel.Session;
+import kafka.security.auth.Operation;
+import kafka.security.auth.Resource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class Auditor {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(Auditor.class);
+
+    protected final ConcurrentHashMap<AuditKey, UserActivity> audit = new ConcurrentHashMap<>();
+
+    private final ScheduledExecutorService auditScheduler =
+        Executors.newScheduledThreadPool(1);
+
+    private final long eventsAggregationPeriod;
+
+    public Auditor(final long eventsAggregationPeriod) {
+        this.eventsAggregationPeriod = eventsAggregationPeriod;
+    }
+
+    public void addActivity(final Session session,
+                            final Operation operation,
+                            final Resource resource,
+                            final Boolean hasAccess) {
+
+        final AuditKey auditKey = new AuditKey(session.principal(), session.clientAddress());
+        if (!audit.containsKey(auditKey)) {
+            audit.computeIfAbsent(
+                auditKey,
+                skip -> {
+                    final UserActivity userActivity = new UserActivity();
+                    return onUserActivity(userActivity, operation, resource, hasAccess);
+                }
+            );
+        } else {
+            audit.computeIfPresent(
+                auditKey,
+                (key, userActivity) -> onUserActivity(userActivity, operation, resource, hasAccess)
+            );
+        }
+    }
+
+    public void start() {
+        auditScheduler.scheduleAtFixedRate(
+            this::dump,
+            eventsAggregationPeriod,
+            eventsAggregationPeriod,
+            TimeUnit.SECONDS
+        );
+    }
+
+    public void stop() {
+        auditScheduler.shutdownNow();
+        dump();
+    }
+
+    protected void dump() {
+        audit.forEach((key, userActivity) -> LOGGER.info("{}", auditMessage(key, userActivity)));
+        audit.clear();
```

Comment:
This is prone to lost records.
Probably we'll have to switch to a normal `HashMap` replacing it with an empty one under lock, like this (a sketch):
```java
    HashMap<AuditKey, UserActivity> oldAudit = null;
    auditLock.lock();
    try {
        oldAudit = audit;
        audit = new HashMap<>();
    } finally {
        auditLock.unlock();
    }
    // dump oldAudit
```

Reply:
it was my concern as well ... i think would be use `synchronized(map) { .... }`